File: solutions/python/VB_T3_004_list_sum.py
"""VB-T3-004: List Sum -- Python baseline."""
class List: pass
class Nil(List): pass
class Cons(List):
    __match_args__ = ("head", "tail")
    def __init__(self, head: int, tail: 'List'):
        self.head = head; self.tail = tail

def list_sum(lst: List) -> int:
    match lst:
        case Nil(): return 0
        case Cons(head, tail): return head + list_sum(tail)

if __name__ == "__main__":
    assert list_sum(Nil()) == 0
    assert list_sum(Cons(1, Cons(2, Cons(3, Nil())))) == 6
    print("VB-T3-004 ok")

File: solutions/typescript/VB_T3_004_list_sum.ts
// VB-T3-004: List Sum -- TypeScript baseline
type List = { tag: "Nil" } | { tag: "Cons"; head: number; tail: List };
function listSum(lst: List): number {
  switch (lst.tag) { case "Nil": return 0; case "Cons": return lst.head + listSum(lst.tail); }
}
console.assert(listSum({ tag: "Cons", head: 1, tail: { tag: "Cons", head: 2, tail: { tag: "Nil" } } }) === 3);
